lib/std/builtin.zig

```diff
@@ -79,11 +79,24 @@ pub const SymbolVisibility = enum {
 /// This data structure is used by the Zig language code generation and
 /// therefore must be kept in sync with the compiler implementation.
 pub const AtomicOrder = enum {
+    /// No specific ordering guarantees are provided for the atomic operation.
     Unordered,
+    /// Guarantees that all sequentially consistent operations that appear in the program order before the atomic operation will be visible to the current thread.
+    /// However, ordering with respect to other concurrent threads is not enforced.
     Monotonic,
+    /// Ensures that all memory accesses and modifications that occur before the atomic operation in the program order are visible
+    /// to the current thread. It guarantees ordering with respect to other threads when they perform a `Release` or `SeqCst` operation on the same atomic value.
     Acquire,
+    /// Guarantees that all memory accesses and modifications that occur after the atomic operation in the program order are visible to
+    /// other threads performing an `Acquire` or `SeqCst` operation on the same atomic value.
     Release,
+    /// Combines the guarantees of both `Acquire` and `Release` orderings:
+    /// ensures that all memory accesses and modifications before the atomic operation are visible to the current thread,
+    /// and all memory accesses and modifications after the atomic operation are visible to other threads.
     AcqRel,
+    /// "Sequentially consistent" - provides the strongest synchronization guarantee among all memory orderings.
+    /// Ensures that all memory accesses and modifications from all threads are seen in a total order that is consistent with the program order.
+    /// This ordering establishes a global order of memory operations and ensures strong consistency across all threads.
     SeqCst,
```

Comment:
This documentation is new and I'm not entirely sure about it because I don't have much experience with threading/atomics overall. CC @kprotty do you think it's accurate?

Reply:
Don't think so:

- Unordered means the operation is atomic but doesnt have to have an observable modification order. At least that's what I gather [from LLVM](https://llvm.org/docs/Atomics.html#unordered:~:text=It%20essentially%20guarantees%20that%20races%20produce%20somewhat%20sane%20results%20instead%20of%20having%20undefined%20behavior) in relation to C++ memory orderings.
- Monotonic enforces an [ordering of reads/writes](https://llvm.org/docs/Atomics.html#unordered:~:text=It%20essentially%20guarantees%20that%20if%20you%20take%20all%20the%20operations%20affecting%20a%20specific%20address%2C%20a%20consistent%20ordering%20exists) on its value to other threads by having a modification order (its not linked to SeqCst). All memory accesses in program order are visible to the current thread even without atomics.
- Acquire ensures that if an atomic value becomes visible through a store, and that store was Release, all memory ops before that Release store are visible to the Acquirer. It doesn't affect other threads which Release.
- SeqCst is not consistent with program order (program order isnt a global/cross-thread concept). AFAIK, it mainly provides a shared modification order across multiple atomic variables. This shared MO doesnt provide happens-before (i.e. "memory ops before become visible" like Acq/Rel) across atomic variables though.

Reply:
Defining the meanings for the orderings might rely on https://github.com/ziglang/zig/issues/6396 being addressed.

Reply:
1. This is missing in the langref, that these result in potential hardware fences and compiler fences (for stuff on the same thread like signal handler and interrupt handler/exceptions are not yet supported.
2. I'm not entirely sure if correct, but I think "happens before" relations and semantics in general are for now according to the C11 memory model with a note that this is planned to be changed eventually (out of thin air problem). This would also belong into the langref.
3. Might be worth to mention somewhere (langref?), that LLVM is currently not able to model and optimize low level primitives at scale and thus intrinsics should be preferred instead of inline assembly. 
Background is that there is no coherent optimization model favored by the verification community, so each backend does its own thing.

Reply:
@kprotty I like your wording so I took over some of what you said into the documentation and removed some that I wasn't sure about. I think this is still a good start and before or after #6396 is addressed we can of course still adjust this further.